server/src/main/java/io/crate/metadata/upgrade/MetadataIndexUpgrader.java

```diff
@@ -90,4 +113,76 @@ MappingMetadata createUpdatedIndexMetadata(MappingMetadata mappingMetadata, Stri
             return mappingMetadata;
         }
     }
+
+    private void reorderPositionsIfNecessary(Map<String, Object> properties) {
+        if (shouldReorder(properties)) {
+            reorderPositions(properties, 1);
+        }
+    }
+
+    private boolean shouldReorder(Map<String, Object> properties) {
+        return isDeep(properties, 0, 2) && containsDuplicatePositions(properties, new HashSet<>());
+    }
+
+    private boolean containsDuplicatePositions(Map<String, Object> properties, Set<Integer> positions) {
+
+        boolean foundDuplicates = false;
+        for (var e : properties.values()) {
+            if (foundDuplicates) {
+                return true;
+            }
+            Map<String, Object> m = (Map<String, Object>) e;
+            Integer position = (Integer) m.get("position");
+            if (position != null) {
+                if (positions.contains(position)) {
+                    return true;
+                } else {
+                    positions.add(position);
+                }
+            }
+            if (m.containsKey("properties")) {
+                foundDuplicates |= containsDuplicatePositions(Maps.get(m, "properties"), positions);
+            }
+        }
+        return foundDuplicates;
+    }
+
+    private boolean isDeep(Map<String, Object> properties, int currentDepth, int targetDepth) {
+        if (++currentDepth >= targetDepth) {
+            return true;
+        }
+        boolean isDeep = false;
+        for (var e : properties.values()) {
+            if (isDeep) {
+                return true;
+            }
+            Map<String, Object> m = (Map<String, Object>) e;
+            if (m.containsKey("properties")) {
+                isDeep |= isDeep(Maps.get(m, "properties"), currentDepth, targetDepth);
+            }
+        }
+        return isDeep;
+    }
+
+    private Integer reorderPositions(Map<String, Object> properties, @Nonnull Integer numFieldsCounter) {
+
+        var sortedProperties = properties.entrySet().stream().sorted(SORT_BY_POSITION_THEN_NAME)
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
+
+        for (var e : sortedProperties.values()) {
+            Map<String, Object> m = (Map<String, Object>) e;
+            Integer position = (Integer) m.get("position");
+            if (position == null) {
+                continue;
+            }
+            if (!position.equals(numFieldsCounter)) {
+                m.replace("position", numFieldsCounter);
+            }
+            numFieldsCounter++;
+            if (m.containsKey("properties")) {
+                numFieldsCounter = reorderPositions(Maps.get(m, "properties"), numFieldsCounter);
```

Comment:
Could replace the `containsKey` / `get` pair with a single `get`. The `null` result is unambiguous in this case.